### ISSUE TITLE
Changing scale messes up rendering for invisible clients using `fractional-scale-v1`

### ISSUE THREAD

**Issue by sysedwinistrator** [swaywm/sway#7762]
### Please fill out the following:
- **Sway Version:**
  - `swaymsg -t get_version` or `sway -v`: `sway version 1.9-dev` (build from master today)

- **Debug Log:**
  - Run `sway -d 2> ~/sway.log` from a TTY and upload it to a pastebin, such as gist.github.com.
  - This will record information about sway's activity. Please try to keep the reproduction as brief as possible and exit sway.
  - Attach the **full** file, do not truncate it.

- **Configuration File:**
  - Please try to produce with the default configuration.
  - If you cannot reproduce with the default configuration, please try to find the minimal configuration to reproduce.
  - Upload the config to a pastebin such as gist.github.com.

- **Description:**
  - Open a window of an application that supports and uses `fractional-scale-v1`
     - Applications tested:
         - Firefox 118.0b7 with config `widget.wayland.fractional-scale.enabled=true`
         - foot 1.15.3
         - calibre 6.17 (seems to be getting support via qtwayland 6.5.2)
  - Make the window invisible by opening another workspace on the same screen, or switching to another window in tabbed layout
  - Change the screen's scale factor via `swaymsg output <output> scale <some fractional scale>`
  - Make the window visible again

While the window will have the correct scale, the rendering will be messed up (jagged fonts etc). Note that this is NOT the blurriness you get when using a fractional scale with applications that don't support the `fractional-scale-v1` protocol. Unfortunately, this doesn't show very well on screenshots taken with grim, but it's very visible in person.  

Meanwhile, the same applications scale fine if they're visible while changing the scale. Even when multiple windows of the same application (Firefox) are running, the visible window scales fine, while the hidden window will have the issue.  
To fix the window, one has to change the scale again (setting the same scale again doesn't work)

I read up a little bit on `fractional-scale-v1` (actually the whole point of compiling sway/wlroots from master and using Firefox Beta was to try out whether I could get fractional scaling to work). Is it possible that sway misses to inform the invisible clients about the new scale via the [wp_fractional_scale_v1:event:preferred_scale](https://wayland.app/protocols/fractional-scale-v1#wp_fractional_scale_v1:event:preferred_scale) event?
 
**EDIT:** I just noticed I had changed the scale_filter to from `linear` to `nearest`. Now with `linear`, it's just blurry instead. However, it's not as visible as the artifacts I get with `linear`. This also makes me wonder: for a client using the `fractional-scale-v1` protocol and scaling its content as the compositor requests, the compositor shouldn't have to scale anything, right? So shouldn't it then make no difference which scale filter is set for the output as no scaling is done?

**Comment by sysedwinistrator** [comment]
I added some debug prints to the code in `sway/desktop/surface.c` that notifies surfaces/clients about the change in scale:
```diff
diff --git a/sway/desktop/surface.c b/sway/desktop/surface.c
index 5932eaa2..352bbfb9 100644
--- a/sway/desktop/surface.c
+++ b/sway/desktop/surface.c
@@ -3,6 +3,7 @@
 #include <time.h>
 #include <wlr/types/wlr_compositor.h>
 #include <wlr/types/wlr_fractional_scale_v1.h>
+#include "log.h"
 #include "sway/server.h"
 #include "sway/surface.h"
 #include "sway/output.h"
@@ -55,6 +56,7 @@ void surface_update_outputs(struct wlr_surface *surface) {
 			scale = surface_output->output->scale;
 		}
 	}
+	sway_log(SWAY_DEBUG, "Notifying surface about new scale");
 	wlr_fractional_scale_v1_notify_scale(surface, scale);
 	wlr_surface_set_preferred_buffer_scale(surface, ceil(scale));
 }
```

When I ran sway with debug loggin, the message was printed only as many times as there were visible clients.

**Comment by emersion** [comment]
> Is it possible that sway misses to inform the invisible clients about the new scale via the [wp_fractional_scale_v1:event:preferred_scale](https://wayland.app/protocols/fractional-scale-v1#wp_fractional_scale_v1:event:preferred_scale) event?

Yeah, sounds like this would happen. We need to broadcast scale events on output scale change.

**Comment by ruabmbua** [comment]
Stupid question: are the monitor related settings (like scale factor) even defined for an invisible client? Since invisible clients are not rendered and not actually placed on any output I suppose broadcasting scale changes to them should be unnecessary.

When the window is "unhidden", the scale factor of the monitor where it pops up should probably be sent as an event to the window.

**Comment by emersion** [comment]
Can you still reproduce this? I can't, invisible tabbed windows get the fractional-scale event.

**Comment by emersion** [comment]
Hm, so I can't reproduce with master, but I can reproduce with the v1.9 branch.

**Comment by stacyharper** [comment]
I dunno about the current master branch. But I keep getting this on 1.9-RC1 yes

**Comment by sysedwinistrator** [comment]
> Can you still reproduce this? I can't, invisible tabbed windows get the fractional-scale event.

You're right, it looks like this has been fixed. :+1: 


> Hm, so I can't reproduce with master, but I can reproduce with the v1.9 branch.

> I dunno about the current master branch. But I keep getting this on 1.9-RC1 yes

I use the nightly builds from [nixpkgs-wayland](https://github.com/nix-community/nixpkgs-wayland).

@emersion feel free to close this issue. For me, as someone who uses nightly builds from the master branch, the issue is fixed; but I don't know whether the fix can be backported to 1.9.
